head_first_design_patterns/tbruinem/json/src/element.cpp

```diff
@@ -0,0 +1,147 @@
+#include "element.hpp"
+#include <assert.h>
+
+#include "common/iostream.hpp"
+
+namespace json {
+
+const ElementType& Element::Type() const
+{
+	return this->type;
+}
+
+void	Element::SetBasicValues(const ParsedJSON& json) {
+	this->type = json.GetType();
+	this->name = json.GetName();
+	this->value_ = json.GetValue();
+}
+
+Element Element::STRING(ParsedJSON json)
+{
+	auto element = Element();
+	element.SetBasicValues(json);
+	assert(element.type == ElementType::JSON_STRING);
+	return move(element);
+}
+
+Element Element::BOOL(ParsedJSON json)
+{
+	auto element = Element();
+	element.SetBasicValues(json);
+	assert(element.type == ElementType::JSON_BOOL);
+	return move(element);
+}
+
+static void FillChildrenVector(vector<unique_ptr<Element>>& children, ParsedJSON* elements) {
+	EmbeddedList* list = elements->GetList();
+	while (list) {
+		auto* parsed = (ParsedJSON*)list->holder;
+		children.push_back(move(unique_ptr<Element>(parsed->Finalize())));
+		list = list->Next();
+		delete(parsed);
+	}
+}
+
+Element Element::ARRAY(ParsedJSON json)
+{
+	auto element = Element();
+	element.SetBasicValues(json);
+	assert(element.type == ElementType::JSON_ARRAY);
+	FillChildrenVector(element.children, (ParsedJSON*)element.value_.pointer);
+	return move(element);
+}
+
+Element Element::GROUP(ParsedJSON json)
+{
+	auto element = Element();
+	element.SetBasicValues(json);
+	assert(element.type == ElementType::JSON_OBJECT);
+	FillChildrenVector(element.children, (ParsedJSON*)element.value_.pointer);
+	return move(element);
+}
+
+Element Element::NULL_ELEM(ParsedJSON json)
+{
+	auto element = Element();
+	element.SetBasicValues(json);
+	assert(element.type == ElementType::JSON_NULL);
+	return move(element);
+}
+
+Element Element::NUMBER(ParsedJSON json)
+{
+	auto element = Element();
+	element.SetBasicValues(json);
+	assert(element.type == ElementType::JSON_INTEGER);
+	return move(element);
+}
+
+static string Indent(size_t depth) {
+	return string(depth, '\t');
+}
+
+static string Name(const string& name) {
+	if (name.empty()) {
+		return "";
+	}
+	else {
+		return "\"" + name + "\": ";
+	}
+}
+
+static string Value(vector<unique_ptr<Element>>& children, const JSONValue& val, ElementType& type, size_t indent_level) {
+	switch (type) {
+		case ElementType::JSON_BOOL: {
+			return val.boolean ? "true" : "false";
+		}
+		case ElementType::JSON_INTEGER: {
+			return std::to_string(val.bigint);
+		}
+		case ElementType::JSON_STRING: {
+			return "\"" + *((string*)(val.pointer)) + "\"";
+		}
+		case ElementType::JSON_NULL: {
+			return "null";
+		}
+		case ElementType::JSON_OBJECT:
+		case ElementType::JSON_ARRAY: {
+			char brackets[2][2] = {
+				{
+					'[',
+					']'
+				},
+				{
+					'{',
+					'}'
+				}
+			};
```

Comment:
How do I print this?

Reply:
Not sure what you mean, Element::ToString ?

Reply:
It was a reference to printing a multi-dimensional array